shell/server/drivers/ip.js

```diff
@@ -37,7 +37,24 @@ ByteStreamConnection.prototype.write = function (data) {
 // expectSize not implemented
 // ByteStreamConnection.prototype.expectSize = function (size) { }
 
-IpInterfaceImpl = function () { };
+function IpInterfaceImpl (userId, parentToken, grainId) {
+  // TODO(someday): maybe check if userId is admin? This could break things if we don't allow
+  // handing off capabilties from a deleted/deactivated user, so for now we're just allowing all
+  // admins to revoke any ipInterface/ipNetwork caps from the admin settings page.
+  this.userId = userId;
+  this.parentToken = parentToken;
+  this.grainId = grainId;
+}
+
+makeIpInterface = function (userId, parentToken, grainId) {
```

Comment:
Tracing through how `grainId` is used, I think it's not being used correctly.

It looks like the `grainId` passed in is the grain that restored the token. It looks like it eventually gets used in `save()` to fill in `grainId` of the `ApiToken`. An ApiToken for a frontend ref should not have a `grainId` at all.

I think you can remove all use of `grainId` in this file and you can also remove it as a parameter to `restoreInternal()`.